File: Serenity.Script.UI/PropertyGrid/EditorTypeRegistry.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [Imported]
    public static class EditorTypeRegistry
    {
        internal static JsDictionary<string, Type> knownTypes;

        public static Type Get(string key)
        {
            if (key.IsEmptyOrNull())
                throw new ArgumentNullException("key");

            Initialize();

            var editorType = knownTypes[key.ToLower()];
            if (editorType == null)
                throw new Exception(String.Format("Can't find {0} editor type!", key));

            return editorType;
        }

        internal static void Initialize()
        {
            if (knownTypes != null)
                return;

            knownTypes = new JsDictionary<string, Type>();
            foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
            {
                foreach (var type in assembly.GetTypes())
                {
                    if (!type.IsSubclassOf(typeof(Widget)))
                        continue;

                    if (type.IsGenericTypeDefinition)
                        continue;

                    var fullName = type.FullName.ToLower();
                    knownTypes[fullName] = type;

                    var editorAttr = type.GetCustomAttributes(typeof(EditorAttribute), false);
                    if (editorAttr != null && editorAttr.Length > 0)
                    {
                        var attrKey = ((EditorAttribute)editorAttr[0]).Key;
                        if (!attrKey.IsEmptyOrNull())
                            knownTypes[attrKey.ToLower()] = type;
                    }
                            
                    foreach (var k in Q.Config.RootNamespaces)
                    {
                        if (fullName.StartsWith(k.ToLower() + "."))
                        {
                            var kx = fullName.Substr(k.Length + 1).ToLower();
                            if (knownTypes[kx] == null)
                                knownTypes[kx] = type;
                        }
                    }
                }
            }

            SetTypeKeysWithoutEditorSuffix();
        }

        public static void Reset()
        {
            knownTypes = null;
        }

        private static void SetTypeKeysWithoutEditorSuffix()
        {
            const string suffix = "editor";

            foreach (var k in Object.Keys(knownTypes))
            {
                if (!k.EndsWith(suffix))
                    continue;
                        
                var p = k.Substr(0, k.Length - suffix.Length);
                if (p.IsEmptyOrNull())
                    continue;

                if (knownTypes[p] != null)
                    continue;

                knownTypes[p] = knownTypes[k];
            }
        }
    }
}

File: src/Serenity.Scripts/CoreLib/Types/EditorTypeRegistry.ts
import { EditorAttribute } from "../Decorators";
import { ArgumentNullException, Config, endsWith, Exception, format, getAttributes, getType, getTypeFullName, getTypes, isEmptyOrNull, startsWith } from "../Q";

// @ts-ignore
let globalObj: any = typeof (global) !== "undefined" ? global : (typeof (window) !== "undefined" ? window : (typeof (self) !== "undefined" ? self : null));

let knownTypes: { [key: string] : any };

export namespace EditorTypeRegistry {
    export function get(key: string): any {

        if (isEmptyOrNull(key)) {
            throw new ArgumentNullException('key');
        }

        initialize();

        var editorType = knownTypes[key.toLowerCase()];
        if (editorType == null) {

            var type = getType(key) ?? getType(key, globalObj);
            if (type != null) {
                knownTypes[key.toLowerCase()] = type as any;
                return type as any;
            }

            throw new Exception(format("Can't find {0} editor type!", key));
        }
        return editorType;

    }

    function initialize(): void {

        if (knownTypes != null)
            return;

        knownTypes = {};

        for (var type of getTypes()) {

            var fullName = getTypeFullName(type).toLowerCase();
            knownTypes[fullName] = type;

            var editorAttr = getAttributes(type, EditorAttribute, false);
            if (editorAttr != null && editorAttr.length > 0) {
                var attrKey = editorAttr[0].key;
                if (!isEmptyOrNull(attrKey)) {
                    knownTypes[attrKey.toLowerCase()] = type;
                }
            }

            for (var k of Config.rootNamespaces) {
                if (startsWith(fullName, k.toLowerCase() + '.')) {
                    var kx = fullName.substr(k.length + 1).toLowerCase();
                    if (knownTypes[kx] == null) {
                        knownTypes[kx] = type;
                    }
                }
            }
        }

        setTypeKeysWithoutEditorSuffix();
    }

    export function reset(): void {
        knownTypes = null;
    }

    function setTypeKeysWithoutEditorSuffix() {
        var keys = Object.keys(knownTypes);
        for (var k of keys) {
            setWithoutSuffix(k, knownTypes[k]);
        }
    }

    function setWithoutSuffix(key: string, t: any) {
        var suffix = 'editor';

        if (!endsWith(key, suffix))
            return;

        var p = key.substr(0, key.length - suffix.length);

        if (isEmptyOrNull(p))
            return;

        if (knownTypes[p] != null)
            return;

        knownTypes[p] = knownTypes[key];
    }
}